// vue3-datepicker/dist/vue3-datepicker.esm.js
import { defineComponent, pushScopeId, popScopeId, openBlock, createBlock, withModifiers, createVNode, renderSlot, resolveDynamicComponent, Fragment, createCommentVNode, renderList, toDisplayString, withScopeId, computed, resolveComponent, withCtx, createTextVNode, ref, watchEffect, withDirectives, vModelText, vShow } from 'vue';
import { isValid, startOfDecade, endOfDecade, eachYearOfInterval, getYear, getDecade, isBefore, isAfter, subYears, addYears, startOfYear, endOfYear, eachMonthOfInterval, isSameMonth, isSameYear, startOfMonth, endOfMonth, startOfWeek, endOfWeek, setDay, eachDayOfInterval, isSameDay, isWithinInterval, startOfDay, endOfDay, subMonths, addMonths, parse, lightFormat } from 'date-fns';
import { formatWithOptions } from 'date-fns/fp';

var script = defineComponent({
    emits: {
        elementClick: (value) => isValid(value),
        left: () => true,
        right: () => true,
        heading: () => true,
    },
    props: {
        headingClickable: {
            type: Boolean,
            default: false,
        },
        leftDisabled: {
            type: Boolean,
            default: false,
        },
        rightDisabled: {
            type: Boolean,
            default: false,
        },
        columnCount: {
            type: Number,
            default: 7,
        },
        items: {
            type: Array,
            default: () => [],
        },
    },
});

const _withId = /*#__PURE__*/withScopeId("data-v-2e128338");

pushScopeId("data-v-2e128338");
const _hoisted_1 = { class: "v3dp__heading" };
const _hoisted_2 = /*#__PURE__*/createVNode("svg", {
  class: "v3dp__heading__icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 6 8"
}, [
  /*#__PURE__*/createVNode("g", {
    fill: "none",
    "fill-rule": "evenodd"
  }, [
    /*#__PURE__*/createVNode("path", {
      stroke: "none",
      d: "M-9 16V-8h24v24z"
    }),
    /*#__PURE__*/createVNode("path", {
      "stroke-linecap": "round",
      "stroke-linejoin": "round",
      d: "M5 0L1 4l4 4"
    })
  ])
], -1 /* HOISTED */);
const _hoisted_3 = /*#__PURE__*/createVNode("svg", {
  class: "v3dp__heading__icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 6 8"
}, [
  /*#__PURE__*/createVNode("g", {
    fill: "none",
    "fill-rule": "evenodd"
  }, [
    /*#__PURE__*/createVNode("path", {
      stroke: "none",
      d: "M15-8v24H-9V-8z"
    }),
    /*#__PURE__*/createVNode("path", {
      "stroke-linecap": "round",
      "stroke-linejoin": "round",
      d: "M1 8l4-4-4-4"
    })
  ])
], -1 /* HOISTED */);
const _hoisted_4 = { class: "v3dp__body" };
const _hoisted_5 = { class: "v3dp__subheading" };
const _hoisted_6 = /*#__PURE__*/createVNode("hr", { class: "v3dp__divider" }, null, -1 /* HOISTED */);
const _hoisted_7 = { class: "v3dp__elements" };
popScopeId();

const render = /*#__PURE__*/_withId(function render(_ctx, _cache, $props, $setup, $data, $options) {
  return (openBlock(), createBlock("div", {
    class: "v3dp__popout",
    style: { '--popout-column-definition': `repeat(${_ctx.columnCount}, 1fr)` },
    onMousedown: _cache[4] || (_cache[4] = withModifiers(() => {}, ["prevent"]))
  }, [
    createVNode("div", _hoisted_1, [
      createVNode("button", {
        class: "v3dp__heading__button",
        disabled: _ctx.leftDisabled,
        onClick: _cache[1] || (_cache[1] = withModifiers($event => (_ctx.$emit('left')), ["stop","prevent"]))
      }, [
        renderSlot(_ctx.$slots, "arrow-left", {}, () => [
          _hoisted_2
        ])
      ], 8 /* PROPS */, ["disabled"]),
      (openBlock(), createBlock(resolveDynamicComponent(_ctx.headingClickable ? 'button' : 'span'), {
        class: "v3dp__heading__center",
        onClick: _cache[2] || (_cache[2] = withModifiers($event => (_ctx.$emit('heading')), ["stop","prevent"]))
      }, {
        default: _withId(() => [
          renderSlot(_ctx.$slots, "heading")
        ]),
        _: 3
      })),
      createVNode("button", {
        class: "v3dp__heading__button",
        disabled: _ctx.rightDisabled,
        onClick: _cache[3] || (_cache[3] = withModifiers($event => (_ctx.$emit('right')), ["stop","prevent"]))
      }, [
        renderSlot(_ctx.$slots, "arrow-right", {}, () => [
          _hoisted_3
        ])
      ], 8 /* PROPS */, ["disabled"])
    ]),
    createVNode("div", _hoisted_4, [
      ('subheading' in _ctx.$slots)
        ? (openBlock(), createBlock(Fragment, { key: 0 }, [
            createVNode("div", _hoisted_5, [
              renderSlot(_ctx.$slots, "subheading")
            ]),
            _hoisted_6
          ], 64 /* STABLE_FRAGMENT */))
        : createCommentVNode("v-if", true),
      createVNode("div", _hoisted_7, [
        renderSlot(_ctx.$slots, "body", {}, () => [
          (openBlock(true), createBlock(Fragment, null, renderList(_ctx.items, (item) => {
            return (openBlock(), createBlock("button", {
              key: item.key,
              disabled: item.disabled,
              class: { selected: item.selected },
              onClick: withModifiers($event => (_ctx.$emit('elementClick', item.value)), ["stop","prevent"])
            }, [
              createVNode("span", null, toDisplayString(item.display), 1 /* TEXT */)
            ], 10 /* CLASS, PROPS */, ["disabled", "onClick"]))
          }), 128 /* KEYED_FRAGMENT */))
        ])
      ])
    ])
  ], 36 /* STYLE, HYDRATE_EVENTS */))
});

script.render = render;
script.__scopeId = "data-v-2e128338";
script.__file = "src/datepicker/PickerPopup.vue";

var script$1 = defineComponent({
    components: {
        PickerPopup: script,
    },
    emits: {
        'update:pageDate': (date) => isValid(date),
        select: (date) => isValid(date),
    },
    props: {
        selected: {
            type: Date,
            required: false,
        },
        pageDate: {
            type: Date,
            required: true,
        },
        lowerLimit: {
            type: Date,
            required: false,
        },
        upperLimit: {
            type: Date,
            required: false,
        },
    },
    setup(props, { emit }) {
        const from = computed(() => startOfDecade(props.pageDate));
        const to = computed(() => endOfDecade(props.pageDate));
        const isEnabled = (target, lower, upper) => {
            if (!lower && !upper)
                return true;
            if (lower && getYear(target) < getYear(lower))
                return false;
            if (upper && getYear(target) > getYear(upper))
                return false;
            return true;
        };
        const years = computed(() => eachYearOfInterval({
            start: from.value,
            end: to.value,
        }).map((value) => ({
            value,
            key: String(getYear(value)),
            display: getYear(value),
            selected: props.selected && getYear(value) === getYear(props.selected),
            disabled: !isEnabled(value, props.lowerLimit, props.upperLimit),
        })));
        const heading = computed(() => {
            const start = getYear(from.value);
            const end = getYear(to.value);
            return `${start} - ${end}`;
        });
        const leftDisabled = computed(() => props.lowerLimit &&
            (getDecade(props.lowerLimit) === getDecade(props.pageDate) ||
                isBefore(props.pageDate, props.lowerLimit)));
        const rightDisabled = computed(() => props.upperLimit &&
            (getDecade(props.upperLimit) === getDecade(props.pageDate) ||
                isAfter(props.pageDate, props.upperLimit)));
        const previousPage = () => emit('update:pageDate', subYears(props.pageDate, 10));
        const nextPage = () => emit('update:pageDate', addYears(props.pageDate, 10));
        return {
            years,
            heading,
            leftDisabled,
            rightDisabled,
            previousPage,
            nextPage,
        };
    },
});

function render$1(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_picker_popup = resolveComponent("picker-popup");

  return (openBlock(), createBlock(_component_picker_popup, {
    columnCount: 3,
    leftDisabled: _ctx.leftDisabled,
    rightDisabled: _ctx.rightDisabled,
    items: _ctx.years,
    onLeft: _ctx.previousPage,
    onRight: _ctx.nextPage,
    onElementClick: _cache[1] || (_cache[1] = $event => (_ctx.$emit('select', $event)))
  }, {
    heading: withCtx(() => [
      createTextVNode(toDisplayString(_ctx.heading), 1 /* TEXT */)
    ]),
    _: 1
  }, 8 /* PROPS */, ["leftDisabled", "rightDisabled", "items", "onLeft", "onRight"]))
}

script$1.render = render$1;
script$1.__file = "src/datepicker/YearPicker.vue";

var script$2 = defineComponent({
    components: {
        PickerPopup: script,
    },
    emits: {
        'update:pageDate': (date) => isValid(date),
        select: (date) => isValid(date),
        back: () => true,
    },
    props: {
        /**
         * Currently selected date, needed for highlighting
         */
        selected: {
            type: Date,
            required: false,
        },
        pageDate: {
            type: Date,
            required: true,
        },
        format: {
            type: String,
            required: false,
            default: 'LLL',
        },
        locale: {
            type: Object,
            required: false,
        },
        lowerLimit: {
            type: Date,
            required: false,
        },
        upperLimit: {
            type: Date,
            required: false,
        },
    },
    setup(props, { emit }) {
        const from = computed(() => startOfYear(props.pageDate));
        const to = computed(() => endOfYear(props.pageDate));
        const format = computed(() => formatWithOptions({
            locale: props.locale,
        })(props.format));
        const isEnabled = (target, lower, upper) => {
            if (!lower && !upper)
                return true;
            if (lower && isBefore(target, startOfMonth(lower)))
                return false;
            if (upper && isAfter(target, endOfMonth(upper)))
                return false;
            return true;
        };
        const months = computed(() => eachMonthOfInterval({
            start: from.value,
            end: to.value,
        }).map((value) => ({
            value,
            display: format.value(value),
            key: format.value(value),
            selected: props.selected && isSameMonth(props.selected, value),
            disabled: !isEnabled(value, props.lowerLimit, props.upperLimit),
        })));
        const heading = computed(() => getYear(from.value));
        const leftDisabled = computed(() => props.lowerLimit &&
            (isSameYear(props.lowerLimit, props.pageDate) ||
                isBefore(props.pageDate, props.lowerLimit)));
        const rightDisabled = computed(() => props.upperLimit &&
            (isSameYear(props.upperLimit, props.pageDate) ||
                isAfter(props.pageDate, props.upperLimit)));
        const previousPage = () => emit('update:pageDate', subYears(props.pageDate, 1));
        const nextPage = () => emit('update:pageDate', addYears(props.pageDate, 1));
        return {
            months,
            heading,
            leftDisabled,
            rightDisabled,
            previousPage,
            nextPage,
        };
    },
});

function render$2(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_picker_popup = resolveComponent("picker-popup");

  return (openBlock(), createBlock(_component_picker_popup, {
    headingClickable: "",
    columnCount: 3,
    items: _ctx.months,
    leftDisabled: _ctx.leftDisabled,
    rightDisabled: _ctx.rightDisabled,
    onLeft: _ctx.previousPage,
    onRight: _ctx.nextPage,
    onHeading: _cache[1] || (_cache[1] = $event => (_ctx.$emit('back'))),
    onElementClick: _cache[2] || (_cache[2] = $event => (_ctx.$emit('select', $event)))
  }, {
    heading: withCtx(() => [
      createTextVNode(toDisplayString(_ctx.heading), 1 /* TEXT */)
    ]),
    _: 1
  }, 8 /* PROPS */, ["items", "leftDisabled", "rightDisabled", "onLeft", "onRight"]))
}

script$2.render = render$2;
script$2.__file = "src/datepicker/MonthPicker.vue";

var script$3 = defineComponent({
    components: {
        PickerPopup: script,
    },
    emits: {
        'update:pageDate': (date) => isValid(date),
        select: (date) => isValid(date),
        back: () => true,
    },
    props: {
        selected: {
            type: Date,
            required: false,
        },
        pageDate: {
            type: Date,
            required: true,
        },
        format: {
            type: String,
            required: false,
            default: 'dd',
        },
        headingFormat: {
            type: String,
            required: false,
            default: 'LLLL yyyy',
        },
        weekdayFormat: {
            type: String,
            required: false,
            default: 'EE',
        },
        locale: {
            type: Object,
            required: false,
        },
        weekStartsOn: {
            type: Number,
            required: false,
            default: 1,
            validator: (i) => typeof i === 'number' && Number.isInteger(i) && i >= 0 && i <= 6,
        },
        lowerLimit: {
            type: Date,
            required: false,
        },
        upperLimit: {
            type: Date,
            required: false,
        },
    },
    setup(props, { emit }) {
        const format = computed(() => formatWithOptions({
            locale: props.locale,
            weekStartsOn: props.weekStartsOn,
        }));
        const monthStart = computed(() => startOfMonth(props.pageDate));
        const monthEnd = computed(() => endOfMonth(props.pageDate));
        const currentMonth = computed(() => ({
            start: monthStart.value,
            end: monthEnd.value,
        }));
        const displayedInterval = computed(() => ({
            start: startOfWeek(monthStart.value, {
                weekStartsOn: props.weekStartsOn,
            }),
            end: endOfWeek(monthEnd.value, {
                weekStartsOn: props.weekStartsOn,
            }),
        }));
        const weekDays = computed(() => {
            const initial = props.weekStartsOn;
            const dayFormat = format.value(props.weekdayFormat);
            return Array.from(Array(7))
                .map((_, i) => (initial + i) % 7)
                .map((v) => setDay(new Date(), v, {
                weekStartsOn: props.weekStartsOn,
            }))
                .map(dayFormat);
        });
        const isEnabled = (target, lower, upper) => {
            if (!lower && !upper)
                return true;
            if (lower && isBefore(target, startOfDay(lower)))
                return false;
            if (upper && isAfter(target, endOfDay(upper)))
                return false;
            return true;
        };
        const days = computed(() => {
            const dayFormat = format.value(props.format);
            return eachDayOfInterval(displayedInterval.value).map((value) => ({
                value,
                display: dayFormat(value),
                selected: props.selected && isSameDay(props.selected, value),
                disabled: !isWithinInterval(value, currentMonth.value) ||
                    !isEnabled(value, props.lowerLimit, props.upperLimit),
                key: format.value('yyyy-MM-dd', value),
            }));
        });
        const heading = computed(() => format.value(props.headingFormat)(props.pageDate));
        const leftDisabled = computed(() => props.lowerLimit &&
            (isSameMonth(props.lowerLimit, props.pageDate) ||
                isBefore(props.pageDate, props.lowerLimit)));
        const rightDisabled = computed(() => props.upperLimit &&
            (isSameMonth(props.upperLimit, props.pageDate) ||
                isAfter(props.pageDate, props.upperLimit)));
        const previousPage = () => emit('update:pageDate', subMonths(props.pageDate, 1));
        const nextPage = () => emit('update:pageDate', addMonths(props.pageDate, 1));
        return {
            weekDays,
            days,
            heading,
            leftDisabled,
            rightDisabled,
            previousPage,
            nextPage,
        };
    },
});

function render$3(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_picker_popup = resolveComponent("picker-popup");

  return (openBlock(), createBlock(_component_picker_popup, {
    headingClickable: "",
    leftDisabled: _ctx.leftDisabled,
    rightDisabled: _ctx.rightDisabled,
    items: _ctx.days,
    onLeft: _ctx.previousPage,
    onRight: _ctx.nextPage,
    onHeading: _cache[1] || (_cache[1] = $event => (_ctx.$emit('back'))),
    onElementClick: _cache[2] || (_cache[2] = $event => (_ctx.$emit('select', $event)))
  }, {
    heading: withCtx(() => [
      createTextVNode(toDisplayString(_ctx.heading), 1 /* TEXT */)
    ]),
    subheading: withCtx(() => [
      (openBlock(true), createBlock(Fragment, null, renderList(_ctx.weekDays, (day) => {
        return (openBlock(), createBlock("span", { key: day }, toDisplayString(day), 1 /* TEXT */))
      }), 128 /* KEYED_FRAGMENT */))
    ]),
    _: 1
  }, 8 /* PROPS */, ["leftDisabled", "rightDisabled", "items", "onLeft", "onRight"]))
}

script$3.render = render$3;
script$3.__file = "src/datepicker/DayPicker.vue";

var script$4 = defineComponent({
    components: {
        YearPicker: script$1,
        MonthPicker: script$2,
        DayPicker: script$3,
    },
    props: {
        placeholder: {
            type: String,
            default: '',
        },
        /**
         * `v-model` for selected date
         */
        modelValue: {
            type: Date,
            required: false,
        },
        /**
         * Upper limit for available dates for picking
         */
        upperLimit: {
            type: Date,
            required: false,
        },
        /**
         * Lower limit for available dates for picking
         */
        lowerLimit: {
            type: Date,
            required: false,
        },
        /**
         * View on which the date picker should open. Can be either `year`, `month`, or `day`
         */
        startingView: {
            type: String,
            required: false,
            default: 'day',
            validate: (v) => typeof v === 'string' && ['day', 'month', 'year'].includes(v),
        },
        /**
         * `date-fns`-type formatting for a month view heading
         */
        monthHeadingFormat: {
            type: String,
            required: false,
            default: 'LLLL yyyy',
        },
        /**
         * `date-fns`-type formatting for the month picker view
         */
        monthListFormat: {
            type: String,
            required: false,
            default: 'LLL',
        },
        /**
         * `date-fns`-type formatting for a line of weekdays on day view
         */
        weekdayFormat: {
            type: String,
            required: false,
            default: 'EE',
        },
        /**
         * `date-fns`-type format in which the string in the input should be both
         * parsed and displayed
         */
        inputFormat: {
            type: String,
            required: false,
            default: 'yyyy-MM-dd',
        },
        /**
         * [`date-fns` locale object](https://date-fns.org/v2.16.1/docs/I18n#usage).
         * Used in string formatting (see default `monthHeadingFormat`)
         */
        locale: {
            type: Object,
            required: false,
        },
        /**
         * Day on which the week should start.
         *
         * Number from 0 to 6, where 0 is Sunday and 6 is Saturday.
         * Week starts with a Monday (1) by default
         */
        weekStartsOn: {
            type: Number,
            required: false,
            default: 1,
            validator: (value) => [0, 1, 2, 3, 4, 5, 6].includes(value),
        },
        /**
         * Disables datepicker and prevents it's opening
         */
        disabled: {
            type: Boolean,
            required: false,
            default: false,
        },
    },
    setup(props, { emit }) {
        const viewShown = ref('none');
        const pageDate = ref(new Date());
        const input = ref('');
        watchEffect(() => {
            const parsed = parse(input.value, props.inputFormat, new Date());
            if (isValid(parsed)) {
                pageDate.value = parsed;
            }
        });
        watchEffect(() => (input.value =
            props.modelValue && isValid(props.modelValue)
                ? lightFormat(props.modelValue, props.inputFormat)
                : ''));
        const renderView = (view = 'none') => {
            if (!props.disabled)
                viewShown.value = view;
        };
        watchEffect(() => {
            if (props.disabled)
                viewShown.value = 'none';
        });
        const selectYear = (date) => {
            pageDate.value = date;
            viewShown.value = 'month';
        };
        const selectMonth = (date) => {
            pageDate.value = date;
            viewShown.value = 'day';
        };
        const selectDay = (date) => {
            emit('update:modelValue', date);
            viewShown.value = 'none';
        };
        return {
            input,
            pageDate,
            renderView,
            selectYear,
            selectMonth,
            selectDay,
            viewShown,
            log: (e) => console.log(e),
        };
    },
});

const _hoisted_1$1 = { class: "v3dp__datepicker" };

function render$4(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_year_picker = resolveComponent("year-picker");
  const _component_month_picker = resolveComponent("month-picker");
  const _component_day_picker = resolveComponent("day-picker");

  return (openBlock(), createBlock("div", _hoisted_1$1, [
    withDirectives(createVNode("input", {
      type: "text",
      readonly: "readonly",
      "onUpdate:modelValue": _cache[1] || (_cache[1] = $event => (_ctx.input = $event)),
      placeholder: _ctx.placeholder,
      disabled: _ctx.disabled,
      tabindex: _ctx.disabled ? -1 : 0,
      onBlur: _cache[2] || (_cache[2] = $event => (_ctx.renderView())),
      onFocus: _cache[3] || (_cache[3] = $event => (_ctx.renderView(_ctx.startingView))),
      onClick: _cache[4] || (_cache[4] = $event => (_ctx.renderView(_ctx.startingView)))
    }, null, 40 /* PROPS, HYDRATE_EVENTS */, ["placeholder", "disabled", "tabindex"]), [
      [vModelText, _ctx.input]
    ]),
    withDirectives(createVNode(_component_year_picker, {
      pageDate: _ctx.pageDate,
      "onUpdate:pageDate": _cache[5] || (_cache[5] = $event => (_ctx.pageDate = $event)),
      selected: _ctx.modelValue,
      lowerLimit: _ctx.lowerLimit,
      upperLimit: _ctx.upperLimit,
      onSelect: _ctx.selectYear
    }, null, 8 /* PROPS */, ["pageDate", "selected", "lowerLimit", "upperLimit", "onSelect"]), [
      [vShow, _ctx.viewShown === 'year']
    ]),
    withDirectives(createVNode(_component_month_picker, {
      pageDate: _ctx.pageDate,
      "onUpdate:pageDate": _cache[6] || (_cache[6] = $event => (_ctx.pageDate = $event)),
      selected: _ctx.modelValue,
      onSelect: _ctx.selectMonth,
      lowerLimit: _ctx.lowerLimit,
      upperLimit: _ctx.upperLimit,
      format: _ctx.monthListFormat,
      headingFormat: _ctx.monthHeadingFormat,
      locale: _ctx.locale,
      onBack: _cache[7] || (_cache[7] = $event => (_ctx.viewShown = 'year'))
    }, null, 8 /* PROPS */, ["pageDate", "selected", "onSelect", "lowerLimit", "upperLimit", "format", "headingFormat", "locale"]), [
      [vShow, _ctx.viewShown === 'month']
    ]),
    withDirectives(createVNode(_component_day_picker, {
      pageDate: _ctx.pageDate,
      "onUpdate:pageDate": _cache[8] || (_cache[8] = $event => (_ctx.pageDate = $event)),
      selected: _ctx.modelValue,
      weekStartsOn: _ctx.weekStartsOn,
      lowerLimit: _ctx.lowerLimit,
      upperLimit: _ctx.upperLimit,
      locale: _ctx.locale,
      weekdayFormat: _ctx.weekdayFormat,
      onSelect: _ctx.selectDay,
      onBack: _cache[9] || (_cache[9] = $event => (_ctx.viewShown = 'month'))
    }, null, 8 /* PROPS */, ["pageDate", "selected", "weekStartsOn", "lowerLimit", "upperLimit", "locale", "weekdayFormat", "onSelect"]), [
      [vShow, _ctx.viewShown === 'day']
    ])
  ]))
}

script$4.render = render$4;
script$4.__file = "src/datepicker/Datepicker.vue";

export default script$4;
